File: base/src/main/java/messyellie/storagecores/base/data/styles/StylesFile.java
package messyellie.storagecores.base.data.styles;

import com.mojang.serialization.Codec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import messyellie.storagecores.base.data.ReplacingMap;
import net.minecraft.resources.ResourceLocation;

import java.util.Map;

public record StylesFile(Map<ResourceLocation, StyleEntry> entries, boolean replace) implements ReplacingMap<StyleEntry> {
    public static final Codec<StylesFile> CODEC = RecordCodecBuilder.create(
            instance -> instance.group(
                            Codec.unboundedMap(ResourceLocation.CODEC, StyleEntry.CODEC).fieldOf("values").forGetter(StylesFile::entries),
                            Codec.BOOL.optionalFieldOf("replace", Boolean.FALSE).forGetter(StylesFile::replace)
                    )
                    .apply(instance, StylesFile::new)
    );

    @Override
    public boolean shouldReplace() {
        return replace;
    }

    @Override
    public Map<ResourceLocation, StyleEntry> values() {
        return entries;
    }
}

File: base/src/main/kotlin/compass_system/storagecores/base/data/styles/StylesFile.kt
package compass_system.storagecores.base.data.styles

import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import compass_system.storagecores.base.data.ReplacingMap
import net.minecraft.resources.ResourceLocation

data class StylesFile(val values: Map<ResourceLocation, StyleEntry>, val replace: Boolean) : ReplacingMap<StyleEntry> {
    override fun replaceExistingEntries() = replace

    override fun entries() = values

    companion object {
        val codec: Codec<StylesFile> = RecordCodecBuilder.create { instance ->
            instance.group(
                Codec.unboundedMap(ResourceLocation.CODEC, StyleEntry.codec).fieldOf("values").forGetter(StylesFile::values),
                Codec.BOOL.optionalFieldOf("replace", false).forGetter(StylesFile::replace)
            ).apply(instance, ::StylesFile)
        }
    }
}
